src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+using Microsoft.Quantum.QsCompiler.TextProcessing;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.Serialization;
+using static Microsoft.Quantum.QsCompiler.TextProcessing.CompletionParsing;
+
+namespace Microsoft.Quantum.QsCompiler.CompilationBuilder
+{
+    /// <summary>
+    /// Data associated with a completion item that is used for resolving additional information.
+    /// </summary>
+    [DataContract]
+    public class CompletionItemData
+    {
+        [DataMember(Name = "namespace")]
+        private string @namespace;
+
+        [DataMember(Name = "name")]
+        private string name;
+
+        /// <summary>
+        /// The text document that the original completion request was made from.
+        /// </summary>
+        [DataMember(Name = "textDocument")]
+        public TextDocumentIdentifier TextDocument { get; set; }
+
+        /// <summary>
+        /// The qualified name of the completion item.
+        /// </summary>
+        public QsQualifiedName QualifiedName
+        {
+            get =>
+                @namespace == null || name == null
+                ? null
+                : new QsQualifiedName(NonNullable<string>.New(@namespace), NonNullable<string>.New(name));
+            set
+            {
+                @namespace = value.Namespace.Value;
+                name = value.Name.Value;
+            }
+        }
+
+        /// <summary>
+        /// The source file the completion item is declared in.
+        /// </summary>
+        [DataMember(Name = "sourceFile")]
+        public string SourceFile { get; set; }
+    }
+
+    /// <summary>
+    /// Provides code completion for the language server.
+    /// </summary>
+    internal static class EditorCompletion
+    {
+        /// <summary>
+        /// Describes the relationship between two code fragments.
+        /// </summary>
+        private enum FragmentRelationship
+        {
+            /// <summary>
+            /// One or both code fragments are missing.
+            /// </summary>
+            Missing,
+
+            /// <summary>
+            /// The code fragments are the same.
+            /// </summary>
+            Self,
+
+            /// <summary>
+            /// Both code fragments are in the same scope.
+            /// </summary>
+            SameScope,
+
+            /// <summary>
+            /// The second fragment is inside an inner scope relative to the first fragment.
+            /// </summary>
+            InnerScope,
+
+            /// <summary>
+            /// The second fragment is in the enclosing scope of the first fragment.
+            /// </summary>
+            EnclosingScope
+        }
+
+        /// <summary>
+        /// Completion items for built-in type keywords.
+        /// </summary>
+        private static readonly IEnumerable<CompletionItem> typeKeywords =
+            new[]
```

Comment:
I removed this list in a newer branch by having the completion parser return all of the type keywords individually instead of just `Type`. I think it would be possible to bring that change to this PR as well, should I do that?

Reply:
The change would only let me get rid of the list of types - the list of characteristics is still there for now but I think I can remove it when I replace the characteristics expression parser later.

Reply:
Sounds good.